online-shop/src/app/data.service.ts

```diff
@@ -0,0 +1,142 @@
+import { HttpClient, HttpErrorResponse } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import { Observable, throwError } from 'rxjs';
+import { IProduct } from './product-list/product-list.component';
+import { catchError } from 'rxjs/operators';
+
+export interface ICartProduct {
+  productId: number,
+  category: string,
+  name: string,
+  price: number,
+  quantity: number,
+}
+
+interface IOrder {
+  customer: string,
+  products: ICartProduct[],
+}
+
+export interface IUser {
+  username: string,
+  password: string | null,
+  fullName?: string,
+  roles?: string[]
+}
+
+
+@Injectable({
+  providedIn: 'root'
+})
+export class DataService {
+  private shoppingCart: ICartProduct[] = [];
+  private BASE_URL = "http://localhost:3000";
```

Comment:
Base URL should be stored in the environments folder (environment.ts file). Base URL can be different for different environments (testing/production...)